File: Queue/queue.go
// Queue Data Structure
/*
	In this implementation, we define a struct for the queue that has a slice to store the values and pointers to the 
	front and rear of the queue. We also define methods for checking if the queue is empty, adding an element to the 
	rear of the queue, and removing an element from the front of the queue.

	The NewQueue() function creates a new queue and initializes its front and rear pointers to nil. The IsEmpty() 
	method checks if the queue is empty by checking if the front pointer is nil. The Enqueue() method adds an 
	element to the rear of the queue by appending it to the slice and updating the rear pointer. The Dequeue() 
	method removes an element from the front of the queue by removing it from the slice and updating the front 
	pointer.

	In the main() function, we create a new queue, add some elements to it, and then dequeue them one 
	by one and print their values.
*/
package main

import "fmt"

// Define a struct for queue that has a slice to store values and pointers to the front and rear of the queue
type Queue struct {
    values []int
    front  *int
    rear   *int
}

// Function to create a new queue and initialize its front and rear pointers
func NewQueue() *Queue {
    q := Queue{values: make([]int, 0), front: nil, rear: nil}
    return &q
}

// Function to check if the queue is empty
func (q *Queue) IsEmpty() bool {
    return q.front == nil
}

// Function to add an element to the rear of the queue
func (q *Queue) Enqueue(val int) {
    // If the queue is empty, set both front and rear pointers to the new element
    if q.front == nil {
        q.front = &val
        q.rear = &val
    } else {
        // Otherwise, add the new element to the rear of the queue and update the rear pointer
        q.values = append(q.values, val)
        q.rear = &q.values[len(q.values)-1]
    }
}

// Function to remove an element from the front of the queue and return its value
func (q *Queue) Dequeue() (int, error) {
    // If the queue is empty, return an error
    if q.front == nil {
        return 0, fmt.Errorf("queue is empty")
    }

    // Get the value of the element at the front of the queue
    val := *q.front

    // If there is only one element in the queue, set both front and rear pointers to nil
    if len(q.values) == 1 {
        q.front = nil
        q.rear = nil
    } else {
        // Otherwise, remove the front element from the slice and update the front pointer
        q.values = q.values[1:]
        q.front = &q.values[0]
    }

    // Return the value of the dequeued element
    return val, nil
}

func main() {
    // Create a new queue
    q := NewQueue()

    // Add some elements to the queue
    q.Enqueue(10)
    q.Enqueue(20)
    q.Enqueue(30)

    // Dequeue elements from the queue and print their values
    for !q.IsEmpty() {
        val, err := q.Dequeue()
        if err != nil {
            fmt.Println(err)
        } else {
            fmt.Println(val)
        }
    }
}

File: Queue/queue.py
# Implement Queue Data Structure
'''
    This implementation uses a Python list to store the items in the queue. The __init__() method initializes 
    an empty list. The is_empty() method checks whether the list is empty or not by checking the length of 
    the list. The enqueue() method adds an item to the back of the queue by appending it to the end of the 
    list. The dequeue() method removes and returns the item at the front of the queue by using the pop() 
    method to remove the first item in the list. If the list is empty, the method raises an IndexError. 
    The peek() method returns the item at the front of the queue without removing it by returning the 
    first item in the list. If the list is empty, the method raises an IndexError. The size() method 
    returns the number of items in the list by returning the length of the list.
'''
class Queue:
    def __init__(self):
        """
        Initializes an empty queue.
        """
        self.items = []

    def is_empty(self):
        """
        Returns True if the queue is empty, False otherwise.
        """
        return len(self.items) == 0

    def enqueue(self, item):
        """
        Adds the given item to the back of the queue.
        """
        self.items.append(item)

    def dequeue(self):
        """
        Removes and returns the item at the front of the queue.
        If the queue is empty, raises an IndexError.
        """
        if self.is_empty():
            raise IndexError("Queue is empty")
        return self.items.pop(0)

    def peek(self):
        """
        Returns the item at the front of the queue without removing it.
        If the queue is empty, raises an IndexError.
        """
        if self.is_empty():
            raise IndexError("Queue is empty")
        return self.items[0]

    def size(self):
        """
        Returns the number of items in the queue.
        """
        return len(self.items)
